csharp/ql/src/semmle/code/csharp/dataflow/SignAnalysis.qll

```diff
@@ -0,0 +1,595 @@
+    comp.getGreaterOperand() = ssaRead(v, 0) and
+    (if comp.isStrict() then isStrict = true else isStrict = false)
+    or
+    testIsTrue = false and
+    comp.getGreaterOperand() = lowerbound and
+    comp.getLesserOperand() = ssaRead(v, 0) and
+    (if comp.isStrict() then isStrict = false else isStrict = true)
+  )
+}
+
+/**
+ * Holds if `upperbound` is an upper bound for `v` at `pos`. This is restricted
+ * to only include bounds for which we might determine a sign.
+ */
+private predicate upperBound(Expr upperbound, Definition v, SsaReadPosition pos, boolean isStrict) {
+  exists(boolean testIsTrue, RelationalOperation comp |
+    pos.hasReadOfVar(v) and
+    guardControlsSsaRead(comp, pos, testIsTrue) and
+    not unknownSign(upperbound)
+  |
+    testIsTrue = true and
+    comp.getGreaterOperand() = upperbound and
+    comp.getLesserOperand() = ssaRead(v, 0) and
+    (if comp.isStrict() then isStrict = true else isStrict = false)
+    or
+    testIsTrue = false and
+    comp.getLesserOperand() = upperbound and
+    comp.getGreaterOperand() = ssaRead(v, 0) and
+    (if comp.isStrict() then isStrict = false else isStrict = true)
+  )
+}
+
+/**
+ * Holds if `eqbound` is an equality/inequality for `v` at `pos`. This is
+ * restricted to only include bounds for which we might determine a sign. The
+ * boolean `isEq` gives the polarity:
+ *  - `isEq = true` : `v = eqbound`
+ *  - `isEq = false` : `v != eqbound`
+ */
+private predicate eqBound(Expr eqbound, Definition v, SsaReadPosition pos, boolean isEq) {
+  exists(Guard guard, boolean testIsTrue, boolean polarity |
+    pos.hasReadOfVar(v) and
+    guardControlsSsaRead(guard, pos, testIsTrue) and
+    guard.isEquality(eqbound, ssaRead(v, 0), polarity) and
+    isEq = polarity.booleanXor(testIsTrue).booleanNot() and
+    not unknownSign(eqbound)
+  )
+}
+
+/**
+ * Holds if `bound` is a bound for `v` at `pos` that needs to be positive in
+ * order for `v` to be positive.
+ */
+private predicate posBound(Expr bound, Definition v, SsaReadPosition pos, boolean isStrict) {
+  lowerBound(bound, v, pos, isStrict)
+  or
+  eqBound(bound, v, pos, true) and isStrict = false
+}
+
+/**
+ * Holds if `bound` is a bound for `v` at `pos` that needs to be negative in
+ * order for `v` to be negative.
+ */
+private predicate negBound(Expr bound, Definition v, SsaReadPosition pos, boolean isStrict) {
+  upperBound(bound, v, pos, isStrict)
+  or
+  eqBound(bound, v, pos, true) and isStrict = false
+}
+
+/**
+ * Holds if `bound` is a bound for `v` at `pos` that can restrict whether `v`
+ * can be zero.
+ */
+private predicate zeroBound(Expr bound, Definition v, SsaReadPosition pos) {
+  lowerBound(bound, v, pos, _) or
+  upperBound(bound, v, pos, _) or
+  eqBound(bound, v, pos, _)
+}
+
+/** Holds if `bound` allows `v` to be positive at `pos`. */
+private predicate posBoundOk(Expr bound, Definition v, SsaReadPosition pos) {
+  posBound(bound, v, pos, _) and TPos() = exprSign(bound)
```

Comment:
This is where we use the `strictness` of the bound. (which is not there in Java)